src/lavinmq/mqtt/broker.cr

```diff
@@ -0,0 +1,111 @@
+require "./client"
+require "./consts"
+require "./exchange"
+require "./protocol"
+require "./session"
+require "./sessions"
+require "./retain_store"
+require "../vhost"
+
+module LavinMQ
+  module MQTT
+    class Broker
+      getter vhost, sessions
+
+      # The `Broker` class acts as an intermediary between the `Server` and MQTT connections.
+      # It is initialized by the `Server` and manages client connections, sessions, and message exchange.
+      # Responsibilities include:
+      # - Handling client connections and disconnections
+      # - Managing client sessions (clean and persistent)
+      # - Publishing messages to the exchange
+      # - Subscribing and unsubscribing clients to/from topics
+      # - Handling the retain store
+      # - Interfacing with the virtual host (vhost) and the exchange to route messages
+      # The `Broker` class helps keep the MQTT client concise and focused on the protocol.
+
+      def initialize(@vhost : VHost, @replicator : Clustering::Replicator)
+        @sessions = Sessions.new(@vhost)
+        @clients = Hash(String, Client).new
+        @retain_store = RetainStore.new(File.join(@vhost.data_dir, "mqtt_reatined_store"), @replicator)
+        @exchange = MQTT::Exchange.new(@vhost, EXCHANGE, @retain_store)
+        @vhost.exchanges[EXCHANGE] = @exchange
+      end
+
+      def session_present?(client_id : String, clean_session) : Bool
+        return false if clean_session
+        session = sessions[client_id]?
+        return false if session.nil? || session.clean_session?
+        true
+      end
+
+      def add_client(socket, connection_info, user, packet)
+        if prev_client = @clients[packet.client_id]?
+          prev_client.close("New client #{connection_info.src} (username=#{packet.username}) connected as #{packet.client_id}")
```

Comment:
could this method call raise? eg. in case of socket errors? should it be rescued?

Reply:
If we get a socket error here, it will be rescued in the connection factory.. but we could be more detailed in that error handling. Right now i think we just log it as a connection error and close the socket.